### PR TITLE
[FLINK-35510][statebackend] Implement basic incremental checkpoint fo…

### PR DESCRIPTION
**Pull request by zoltar9264**
…r ForStStateBackend

## What is the purpose of the change

Use low DB api implement a basic incremental checkpoint for ForStStatebackend, follow steps:

1. db.disableFileDeletions()
2. db.getLiveFiles(true)
3. db.entableFileDeletes(false)

## Brief change log

  - Introduce ForStStateDataTransfer
  - Introduce ForStIncrementalSnapshotStrategy

## Verifying this change

This change added tests and can be verified as follows:

  - Added test ForStStateDataTransferTest
  - Added test ForStIncrementalSnapshotStrategyTest

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (no)
  - The S3 file system connector: (no)

### CODE REVIEW

**Comment by zoltar9264** [comment]
@flinkbot run azure

**Comment by zoltar9264** [comment]
@flinkbot run azure

**Comment by zoltar9264** [comment]
@flinkbot run azure

**Comment by zoltar9264** [comment]
@flinkbot run azure

File: flink-state-backends/flink-statebackend-forst/src/main/java/org/apache/flink/state/forst/snapshot/ForStIncrementalSnapshotStrategy.java [hunk 1]
```diff
@@ -0,0 +1,435 @@
...
+
+                if (fileName.startsWith(MANIFEST_FILE_PREFIX)) {
+                    Preconditions.checkState(
+                            toTransferManifestFile == null,
+                            String.format("Backend:%s should only one manifest file.", backendUID));
+                    toTransferManifestFile = filePath;
+                } else if (fileName.endsWith(SST_FILE_SUFFIX)) {
+                    Optional<StreamStateHandle> uploaded =
+                            snapshotResources.previousSnapshot.getUploaded(fileName);
+                    if (uploaded.isPresent()) {
```
Comment (Zakelly) [inline flink-state-backends/flink-statebackend-forst/src/main/java/org/apache/flink/state/forst/snapshot/ForStIncrementalSnapshotStrategy.java:418]: Maybe you should check `checkpointStreamFactory.couldReuseStateHandle(uploaded.get())` as well. Please see `RocksIncrementalSnapshotStrategy#createUploadFilePaths`.

**Review by Zakelly — CHANGES_REQUESTED** [review]
Thanks for the PR, overall LGTM % several comments.

File: flink-state-backends/flink-statebackend-forst/src/main/java/org/apache/flink/state/forst/ForStStateDataTransfer.java [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+
+            FileSystem sourceFilesystem = filePath.getFileSystem();
+            inputStream = sourceFilesystem.open(filePath, READ_BUFFER_SIZE);
+            closeableRegistry.registerCloseable(inputStream);
+
+            outputStream = checkpointStreamFactory.createCheckpointStateOutputStream(stateScope);
+            closeableRegistry.registerCloseable(outputStream);
+
+            while (transferBytes >= 0) {
+                int numBytes = inputStream.read(buffer);
```
[the remark this answers is not in the document]
Comment (zoltar9264) [reply flink-state-backends/flink-statebackend-forst/src/main/java/org/apache/flink/state/forst/ForStStateDataTransfer.java]: Good idea, done.

File: flink-state-backends/flink-statebackend-forst/src/main/java/org/apache/flink/state/forst/snapshot/ForStIncrementalSnapshotStrategy.java (hunk 1 above)
Comment (zoltar9264) [reply flink-state-backends/flink-statebackend-forst/src/main/java/org/apache/flink/state/forst/snapshot/ForStIncrementalSnapshotStrategy.java]: Thanks for remainder !

File: flink-state-backends/flink-statebackend-forst/src/main/java/org/apache/flink/state/forst/snapshot/ForStSnapshotStrategyBase.java [hunk 3]
```diff
@@ -0,0 +1,344 @@
...
+                    } catch (RocksDBException e) {
+                        LOG.error(
+                                "Enable file deletion failed, backendUID:{}, checkpointId:{}.",
+                                backendUID,
+                                checkpointId,
+                                e);
+                        // TODO: 2024/5/28 wangfeifan - Add more robust exception handling
+                    }
+                });
+    }
```
[the remark this answers is not in the document]
Comment (zoltar9264) [reply flink-state-backends/flink-statebackend-forst/src/main/java/org/apache/flink/state/forst/snapshot/ForStSnapshotStrategyBase.java:176]: If call enableFileDeletions after any exception thrown in disableFileDeletions, consider parallel checkpointing, file deletion may be allowed prematurely. WDYT ?

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/v2/RegisteredKeyValueStateBackendMetaInfo.java [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+import java.util.Collections;
+import java.util.Map;
+
+/**
+ * Compound meta information for a registered state in a keyed state backend. This combines all
+ * serializers and the state name.
+ *
+ * @param <S> Type of state value
+ */
+public class RegisteredKeyValueStateBackendMetaInfo<S> extends RegisteredStateMetaInfoBase {
```
[the remark this answers is not in the document]
Comment (zoltar9264) [reply flink-runtime/src/main/java/org/apache/flink/runtime/state/v2/RegisteredKeyValueStateBackendMetaInfo.java:41]: Thanks for help review ! Another reason for creating a new RegisteredKeyValueStateBackendMetaInfo is different StateDescriptor inside, is there some method bypass this problem ?

File: flink-state-backends/flink-statebackend-forst/src/main/java/org/apache/flink/state/forst/snapshot/ForStIncrementalSnapshotStrategy.java [hunk 5]
```diff
@@ -0,0 +1,435 @@
...
+                            stateScope,
+                            snapshotCloseableRegistry,
+                            tmpResourcesRegistry);
+            metaHandles.addAll(miscFilesTransferResult);
+            transferBytes +=
+                    miscFilesTransferResult.stream()
+                            .mapToLong(HandleAndLocalPath::getStateSize)
+                            .sum();
+
+            HandleAndLocalPath manifestFileTransferResult =
```
[the remark this answers is not in the document]
Comment (zoltar9264) [reply flink-state-backends/flink-statebackend-forst/src/main/java/org/apache/flink/state/forst/snapshot/ForStIncrementalSnapshotStrategy.java:365]: Thanks for reminder, I added a todo.

**Comment by zoltar9264** [comment]
@flinkbot run azure

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/v2/RegisteredKeyValueStateBackendMetaInfo.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (zoltar9264) [reply flink-runtime/src/main/java/org/apache/flink/runtime/state/v2/RegisteredKeyValueStateBackendMetaInfo.java]: Consider different StateDescriptor inside, keep the new RegisteredKeyValueStateBackendMetaInfo and add namespace into it.

File: flink-state-backends/flink-statebackend-forst/src/main/java/org/apache/flink/state/forst/snapshot/ForStSnapshotStrategyBase.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (zoltar9264) [reply flink-state-backends/flink-statebackend-forst/src/main/java/org/apache/flink/state/forst/snapshot/ForStSnapshotStrategyBase.java:184]: Sorry, I understand what you mean and done.

**Comment by zoltar9264** [comment]
@flinkbot run azure

**Comment by zoltar9264** [comment]
@flinkbot run azure

**Comment by zoltar9264** [comment]
@flinkbot run azure

**Comment by zoltar9264** [comment]
@flinkbot run azure

**Comment by zoltar9264** [comment]
@flinkbot run azure

File: flink-state-backends/flink-statebackend-forst/src/main/java/org/apache/flink/state/forst/snapshot/ForStIncrementalSnapshotStrategy.java [hunk 6]
```diff
@@ -0,0 +1,450 @@
...
+            List<HandleAndLocalPath> transferredSstHandles = new ArrayList<>(totalFileNum);
+
+            List<Path> toTransferSstFiles = new ArrayList<>(totalFileNum);
+            List<Path> toTransferMiscFiles = new ArrayList<>(totalFileNum);
+            Path toTransferManifestFile = null;
+
+            for (Path filePath : snapshotResources.liveFiles) {
+                final String fileName = filePath.getName();
+
+                if (fileName.startsWith(MANIFEST_FILE_PREFIX)) {
```
Comment (Zakelly) [inline flink-state-backends/flink-statebackend-forst/src/main/java/org/apache/flink/state/forst/snapshot/ForStIncrementalSnapshotStrategy.java:424]: Shall we use:
```suggestion
                if (fileName.equals(snapshotResources.manifestFileName)) {
```
?

**Review by Zakelly — APPROVED** [review]
Overall LGTM

**Comment by zoltar9264** [comment]
@flinkbot run azure

**Comment by zoltar9264** [comment]
@flinkbot run azure

**Comment by Zakelly** [comment]
CI green, merging...

**Pull request merged**